src/class/usbtmc/usbtmc_device.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * usbtmc_device.h
+ *
+ *  Created on: Sep 10, 2019
+ *      Author: nconrad
+ */
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 N Conrad
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * This file is part of the TinyUSB stack.
+ */
+
+
+#ifndef CLASS_USBTMC_USBTMC_DEVICE_H_
+#define CLASS_USBTMC_USBTMC_DEVICE_H_
+
+#include "usbtmc.h"
+
+// Enable 488 mode by default
+#if !defined(CFG_USBTMC_CFG_ENABLE_488)
+#define CFG_CFG_USBTMC_CFG_ENABLE_488 (1)
+#endif
+
+// USB spec says that full-speed must be 8,16,32, or 64.
+// However, this driver implementation requires it to be >=32
+#define USBTMCD_MAX_PACKET_SIZE (64u)
+
+/***********************************************
+ *  Functions to be implemeted by the class implementation
+ */
+
+#if (CFG_USBTMC_CFG_ENABLE_488)
+extern usbtmc_response_capabilities_488_t const tud_usbtmc_app_capabilities;
+#else
+extern usbtmc_response_capabilities_t const tud_usbtmc_app_capabilities;
+#endif
+
+// In order to proceed, app must call call usbtmcd_start_bus_read(rhport) during or soon after:
+// * tud_usbtmc_app_open_cb
+// * tud_usbtmc_app_msg_data_cb
+// * tud_usbtmc_app_msgBulkIn_complete_cb
+// * tud_usbtmc_app_msg_trigger_cb
+// * (successful) tud_usbtmc_app_check_abort_bulk_out_cb
+// * (successful) tud_usbtmc_app_check_abort_bulk_in_cb
+// * (successful) usmtmcd_app_bulkOut_clearFeature_cb
+
+void tud_usbtmc_app_open_cb(uint8_t interface_id);
```

Comment:
should we replace `tud_usbtmc_app` by simply `tud_usbtmc_` dropping **app* in API

Reply:
I'm not sure. I've found it hard to keep track of which functions are for which direction of communication, and keeping app there makes it more clear that certain functions are for implementation by the app, not the class library.

I think other libraries fix this by using structs of function pointers to create a namespace, but the TUSB code at the moment doesn't do that.

I can go either way with this question.

Reply:
It is rather simple, if the API is invoked by the stack and application have to implement that function, it is callback and must end with `_cb()` for **callback**. Otherwise, it doesn't has `_cb()` e.g

tud_cdc_write() : implemented by stack
tud_cdc_rx_cb(): implemented by application since it is a callback from stack.

Reply:
I'm going to change this to be as you suggested.

I usually read left-to-right, so looking for the _cb is less easy, as I need to read left to right to center.

However, there are more levels of complexity....
1. app calls class (tud_class_*)
2. class calls app (tud_class_*_cb)
3. stack calls class (tud_class_* or tud_class_*_cb); // Sometimes with _cb, other times without _cb?
3. class calls stack (usbd_*)

Number 1 and 2 and 3 have overlap. This somewhat impacts the app developer since auto-complete will suggest both.